## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

def currency_format_millions(amount):
    """
    Converts a dollar amount to millions and formats it with a dollar sign and 'M' for millions.
    
    Args:
    amount (float or int): The dollar amount to be converted and formatted.
    
    Returns:
    str: The formatted dollar amount in millions with a dollar sign.
    """
    millions = round(amount / 1_000_000, 2)  # Convert to millions
    return f"${millions}M"  # Format as currency with 'M' for millions

df = pd.read_csv('client_dataset.csv')
df.head()

ModuleNotFoundError: No module named 'pandas'

In [129]:
# View the column names in the data
df.columns.to_list()

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [130]:
# Use the describe function to gather some basic statistics
df.describe()

client_id     order_id  order_week  order_year  unit_price  unit_cost  \
count  54,639.00    54,639.00   54,639.00   54,639.00   54,639.00  54,639.00   
mean   54,837.87 5,470,190.03       11.36    2,022.99      136.27      99.45   
std    25,487.44 2,599,806.93        7.02        0.08      183.87     133.16   
min    10,033.00 1,000,886.00        1.00    2,022.00        0.01       0.01   
25%    33,593.00 3,196,372.00        6.00    2,023.00       20.80      14.84   
50%    53,305.00 5,496,966.00       11.00    2,023.00       68.31      49.89   
75%    78,498.00 7,733,869.00       17.00    2,023.00      173.16     125.57   
max    99,984.00 9,998,480.00       52.00    2,023.00    1,396.23     846.27   

       unit_weight          qty  line_number  
count    54,639.00    54,639.00    54,639.00  
mean          5.00       570.26         2.98  
std           5.33    18,795.52         2.44  
min           0.00         0.00         0.00  
25%           1.45        32.00         1.00  
50%           3.24        68.00         3.00  
75%           6.89       170.00         5.00  
max          46.43 3,958,244.00         9.00

In [131]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [132]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [133]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
filtered_df = df[df['category'] == top_category]
top_subcategory = filtered_df['subcategory'].value_counts().idxmax()
top_subcategory_count = filtered_df['subcategory'].value_counts().max()
print(f"Top Category: {top_category}")
print(f"Top Subcategory: {top_subcategory} with {top_subcategory_count} entries")


Top Category: consumables
Top Subcategory: bathroom supplies with 6424 entries


In [134]:
# Which five clients had the most entries in the data?
top_client_ids = df['client_id'].value_counts().head(5).index.to_list()
top_clients = df[df['client_id'].isin(top_client_ids)]
for client_id in top_client_ids:
    client_info = top_clients[top_clients['client_id'] == client_id].iloc[0]
    full_name = f"{client_info['first']} {client_info['last']}"
    print(f"Client ID: {client_id}, Full Name: {full_name}")


Client ID: 33615, Full Name: Jessica Reyes
Client ID: 66037, Full Name: Angela Everett
Client ID: 46820, Full Name: Bryan Myers
Client ID: 38378, Full Name: Alexandra Young
Client ID: 24741, Full Name: Kendra Garrett


In [135]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = df['client_id'].value_counts().head(5).index.to_list()

# Print or store the top client ids
print("Top 5 Client IDs:", top_client_ids)


Top 5 Client IDs: [33615, 66037, 46820, 38378, 24741]


In [136]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = df['client_id'].value_counts().idxmax()
top_client_data = df[df['client_id'] == top_client_id]
total_units = top_client_data['qty'].sum()
print(f"Total units ordered by client {top_client_id}: {total_units}")



Total units ordered by client 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [137]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']

#Valadating that the subtotal was created
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0    1,096.80     762.71         7.50  105            1 115,164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3   1,056.18  
4      195.10     108.17        46.43   20            1   3,902.00

In [138]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Assuming unit_weight is also a column
df['total_weight'] = df['unit_weight'] * df['qty']

# Applying the shipping price logic
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)

# Display the updated DataFrame
df[['client_id', 'qty', 'unit_weight', 'total_weight', 'shipping_price']]




client_id   qty  unit_weight  total_weight  shipping_price
0          58515   105         7.50        787.50        5,512.50
1          37609    21         1.49         31.29          312.90
2          57113    39         1.68         65.52          458.64
3          46554    29         1.23         35.67          356.70
4          92089    20        46.43        928.60        6,500.20
...          ...   ...          ...           ...             ...
54634      29043    33         2.25         74.25          519.75
54635      41908    47        11.70        549.90        3,849.30
54636      35176   475         4.16      1,976.00       13,832.00
54637      24485   112        18.04      2,020.48       14,143.36
54638      24007  1031         2.07      2,134.17       14,939.19

[54639 rows x 5 columns]

In [139]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
# Calculate the total price and round it to 2 decimal places

df['total_price'] = ((df['subtotal'] + df['shipping_price']) * 1.0925).round(2)

# Display the updated DataFrame with total price in currency format
df[['client_id', 'qty', 'unit_price', 'subtotal', 'shipping_price', 'total_price']]



client_id   qty  unit_price   subtotal  shipping_price  total_price
0          58515   105    1,096.80 115,164.00        5,512.50   131,839.08
1          37609    21       24.95     523.95          312.90       914.26
2          57113    39       13.52     527.28          458.64     1,077.12
3          46554    29       36.42   1,056.18          356.70     1,543.57
4          92089    20      195.10   3,902.00        6,500.20    11,364.40
...          ...   ...         ...        ...             ...          ...
54634      29043    33       83.13   2,743.29          519.75     3,564.87
54635      41908    47      206.59   9,709.73        3,849.30    14,813.24
54636      35176   475       65.66  31,188.50       13,832.00    49,184.90
54637      24485   112        1.48     165.76       14,143.36    15,632.71
54638      24007  1031        3.01   3,103.31       14,939.19    19,711.43

[54639 rows x 6 columns]

In [140]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
# Create the cost column by multiplying unit_cost by qty and adding the shipping_price
df['cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Display the updated DataFrame
df[['client_id', 'qty', 'unit_cost', 'shipping_price', 'cost']]



client_id   qty  unit_cost  shipping_price      cost
0          58515   105     762.71        5,512.50 85,597.05
1          37609    21      15.09          312.90    629.79
2          57113    39       7.86          458.64    765.18
3          46554    29      24.85          356.70  1,077.35
4          92089    20     108.17        6,500.20  8,663.60
...          ...   ...        ...             ...       ...
54634      29043    33      51.60          519.75  2,222.55
54635      41908    47     175.46        3,849.30 12,095.92
54636      35176   475      57.31       13,832.00 41,054.25
54637      24485   112       1.22       14,143.36 14,280.00
54638      24007  1031       1.61       14,939.19 16,599.10

[54639 rows x 5 columns]

In [141]:
# Create a column for the profit of each line using line cost and line price
# Create the profit column by subtracting cost from total price
df['profit'] = df['total_price'] - df['cost']

# Display the updated DataFrame with profit
df[['client_id', 'qty', 'unit_price', 'total_price', 'cost', 'profit']]



client_id   qty  unit_price  total_price      cost    profit
0          58515   105    1,096.80   131,839.08 85,597.05 46,242.03
1          37609    21       24.95       914.26    629.79    284.47
2          57113    39       13.52     1,077.12    765.18    311.94
3          46554    29       36.42     1,543.57  1,077.35    466.22
4          92089    20      195.10    11,364.40  8,663.60  2,700.80
...          ...   ...         ...          ...       ...       ...
54634      29043    33       83.13     3,564.87  2,222.55  1,342.32
54635      41908    47      206.59    14,813.24 12,095.92  2,717.32
54636      35176   475       65.66    49,184.90 41,054.25  8,130.65
54637      24485   112        1.48    15,632.71 14,280.00  1,352.71
54638      24007  1031        3.01    19,711.43 16,599.10  3,112.33

[54639 rows x 6 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [142]:
# Check your work using the totals above
# List of specific Order IDs to look up
order_ids = [2742071, 2173913, 6128929]

# Filter the DataFrame to include only the rows with the specific order id
filtered_orders = df[df['order_id'].isin(order_ids)]

# Grouping to sum the line items for each order
order_totals = filtered_orders.groupby('order_id')['total_price'].sum().reset_index()

# Display the results
print(order_totals)


   order_id  total_price
0   2173913   162,388.71
1   2742071   152,811.89
2   6128929   923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [143]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Note: the instructions are a little confusing here because when I built the top client list above it was based on the five clients that had the most entries in the data as instructed, not by the quanity spend. I used the user list from part 1 here.
top_clients_data = df[df['client_id'].isin(top_client_ids)]
client_order_totals = top_clients_data.groupby('client_id')['total_price'].sum().reset_index()
client_order_totals['total_price'] = client_order_totals['total_price'].apply(lambda x: f"${x:,.2f}")
print(client_order_totals[['client_id', 'total_price']])



   client_id     total_price
0      24741  $82,268,892.02
1      33615   $8,377,308.52
2      38378  $12,906,550.87
3      46820   $9,743,794.36
4      66037  $10,259,514.79


In [144]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = top_clients_data.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()

print(summary_df)

   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862          5,126,448.37  82,268,892.02   
1      33615                  64313          1,828,984.89   8,377,308.52   
2      38378                  73667          3,429,455.40  12,906,550.87   
3      46820                  75768          1,601,448.84   9,743,794.36   
4      66037                  43018          1,395,151.85  10,259,514.79   

   total_profit  
0 36,579,992.31  
1  2,201,994.61  
2  3,271,829.89  
3  2,736,602.72  
4  3,255,031.81  


In [145]:
# Format the data and rename the columns to names suitable for presentation.

summary_df.rename(columns={
    'client_id': 'Client ID',
    'total_units_purchased': 'Units Purchased',
    'total_shipping_price': 'Total Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit'
}, inplace=True)

# Apply the currency_format_millions function to the money columns only
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Display the formatted summary DataFrame
print(summary_df)



   Client ID  Units Purchased Total Shipping Price Total Revenue Total Profit
0      24741           239862               $5.13M       $82.27M      $36.58M
1      33615            64313               $1.83M        $8.38M        $2.2M
2      38378            73667               $3.43M       $12.91M       $3.27M
3      46820            75768                $1.6M        $9.74M       $2.74M
4      66037            43018                $1.4M       $10.26M       $3.26M


In [147]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit', ascending=False, key=lambda x: x.str.replace('$','').str.replace('M','').astype(float))

# Display the sorted DataFrame
print(sorted_summary_df)

   Client ID  Units Purchased Total Shipping Price Total Revenue Total Profit
0      24741           239862               $5.13M       $82.27M      $36.58M
2      38378            73667               $3.43M       $12.91M       $3.27M
4      66037            43018                $1.4M       $10.26M       $3.26M
3      46820            75768                $1.6M        $9.74M       $2.74M
1      33615            64313               $1.83M        $8.38M        $2.2M


In [ ]:
# Summary of findings: (1) The shipping costs are quite high for these orders. (2) There is a big gap in profit between the most profitable client and the 2 - 4th best clients and this company should make sure they keep this client. (3) They did not consider the shipping distance in the shipping costs.